Insert real time thread by ID

timpalpant opened the pull request:
This allows the user to register another thread for real time profiling by passing its identity (thread id), rather than needing to call `insert_real_time_thread()` from within each thread.

Fixes #201. Depends on / includes #204 to fix the real_time_threaded test.

arigo commented:
The PyArg_ParseTuple() need some tweaking:

* You must initialize the local ``thread_id`` variable you pass as optional argument, or make that argument not optional, because right now if we call the function with no argument then we get an undefined value in the local ``thread_id`` variable.

* Using the argument 'k' is arbitrary, as we don't know the size of ``pthread_t`` is the same as the size expected by 'k'.  You need to pass a different variable of the correct type for 'k', and then cast that into a ``pthread_t``, which ideally should be done carefully (see CPython for how to carefully cast between integers and ``pthread_t`` variables)

timpalpant commented:
Thanks @arigo, I appreciate it. I've amended the change to try to address these issues. The first is an obvious blunder, and I've updated to initialize the variable.

The second is a bit more intricate as you note. I've updated to safely cast to `pthread_t` in the same manner as [`PyThread_start_new_thread`](https://github.com/python/cpython/blob/master/Python/thread_pthread.h#L299-L304), taking into account the size of `pthread_t`. (Interestingly it seems to be a plain cast in [`PyThread_get_thread_ident`](https://github.com/python/cpython/blob/master/Python/thread_pthread.h#L319) below).

Apologies if this is different than what you had in mind. I also consulted with the standard library implementation of [`signal.pthread_kill(thread_id)`](https://github.com/python/cpython/blob/master/Modules/signalmodule.c#L1236) that has a similar interface, but it seems to do a naive cast from `unsigned long` to `pthread_t`.

arigo commented:
Looks good to me.  I'm not really contributing to vmprof, but I don't know if anyone is at the moment, so I'll go ahead and merge your pull request.  Sorry if I'm stepping on someone else's toes.
